creator/referral_tokens/schema.py

```diff
@@ -0,0 +1,226 @@
+from creator.events.models import Event
+from creator.users.schema import UserNode
+
+
+class ReferralTokenNode(DjangoObjectType):
+    is_valid = graphene.Boolean(source="is_valid")
+
+    class Meta:
+        model = ReferralToken
+        interfaces = (relay.Node,)
+        filter_fields = ()
+
+    @classmethod
+    def get_node(cls, info, uuid):
+        """
+        Only return if the user is allowed to view referral token
+        """
+        user = info.context.user
+
+        if not (user.has_perm("referral_tokens.view_referraltoken")):
+            raise GraphQLError("Not allowed")
+
+        if user.has_perm("referral_tokens.view_referraltoken"):
+            try:
+                return cls._meta.model.objects.get(uuid=uuid)
+            except cls._meta.model.DoesNotExist:
+                raise GraphQLError("Referral token not found")
+
+        return ReferralToken.objects.none()
+
+
+class ReferralTokenFilter(django_filters.FilterSet):
+    class Meta:
+        model = ReferralToken
+        fields = {
+            "email": ["exact", "contains"],
+            "claimed": ["exact"],
+            "studies": ["exact"],
+            "groups": ["exact"],
+            "created_at": ["gt", "lt"],
+        }
+
+    order_by = OrderingFilter(fields=("created_at",))
+
+
+def get_studies(input):
+    """
+    Convert a list of id of studies into a list of Study objects
+    """
+    studies_input = input.get("studies", [])
+    study_ids = [from_global_id(s)[1] for s in studies_input]
+    studies = Study.objects.filter(kf_id__in=study_ids).all()
+    if len(studies) < len(studies_input):
+        raise GraphQLError("Study does not exist.")
+
+    return studies
+
+
+def get_groups(input):
+    """
+    Convert a list of id of groups into a list of Group objects
+    """
+    groups_input = input.get("groups", [])
+    group_ids = [from_global_id(g)[1] for g in groups_input]
+    groups = Group.objects.filter(id__in=group_ids).all()
+    if len(groups) < len(groups_input):
+        raise GraphQLError("Group does not exist.")
+
+    return groups
+
+
+class ReferralTokenInput(graphene.InputObjectType):
+    email = graphene.String(
+        required=True, description="The email address to send the referral to",
+    )
+
+    studies = graphene.List(
+        graphene.ID,
+        description="A list of studies that the user will be added to upon"
+        " exchange of the token",
+        required=False,
+    )
+
+    groups = graphene.List(
+        graphene.ID,
+        required=False,
+        description="The permission group the user will be given after"
+        " exchanging the token",
+    )
+
+
+class CreateReferralTokenMutation(graphene.Mutation):
+    class Arguments:
+        input = ReferralTokenInput(
+            required=True, description="Attributes for the new referral token"
+        )
+
+    referral_token = Field(ReferralTokenNode)
+
+    def mutate(self, info, input):
+        """
+        Create a new referral token
```

Comment:
Can you document some about how it checks for existing tokens that are still valid for that email and won't send if it finds one?